vault/resource_managed_keys.go

```diff
@@ -0,0 +1,690 @@
+
+	return d.Get(stateKey).(string)
+}
+
+func getManagedKeysConfigData(config map[string]interface{}, sm schemaMap) (string, map[string]interface{}) {
+	data := map[string]interface{}{}
+	var name string
+
+	for blockKey := range sm {
+		if v, ok := config[blockKey]; ok {
+			data[blockKey] = v
+
+			if blockKey == consts.FieldName {
+				name = v.(string)
+			}
+		}
+	}
+
+	return name, data
+}
+
+func getManagedKeysPath(keyType, name string) string {
+	return fmt.Sprintf("sys/managed-keys/%s/%s", keyType, name)
+}
+
+func writeManagedKeysData(d *schema.ResourceData, meta interface{}, providerType string) diag.Diagnostics {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return diag.FromErr(e)
+	}
+
+	var keyType string
+	handlers := map[string]func() schemaMap{
+		kmsTypeAWS:   managedKeysAWSConfigSchema,
+		kmsTypePKCS:  managedKeysPKCSConfigSchema,
+		kmsTypeAzure: managedKeysAzureConfigSchema,
+	}
+
+	switch providerType {
+	case consts.FieldAWS:
+		keyType = kmsTypeAWS
+
+	case consts.FieldPKCS:
+		keyType = kmsTypePKCS
+
+	case consts.FieldAzure:
+		keyType = kmsTypeAzure
+	}
+
+	if mKeysBlocks, ok := d.GetOk(providerType); ok {
+		set := map[string]bool{}
+
+		// populate set of blocks
+		for _, block := range mKeysBlocks.(*schema.Set).List() {
+			m := block.(map[string]interface{})
+			name := m[consts.FieldName].(string)
+			set[name] = true
+		}
+
+		blocksWritten := map[string]bool{}
+		for _, block := range mKeysBlocks.(*schema.Set).List() {
+			keyName, data := getManagedKeysConfigData(block.(map[string]interface{}), handlers[keyType]())
+			path := getManagedKeysPath(keyType, keyName)
+
+			log.Printf("[DEBUG] Writing data to Vault at %s", path)
+			if _, err := client.Logical().Write(path, data); err != nil {
+				return diag.Errorf("error writing managed key %q, err=%s", path, err)
+			}
+
+			blocksWritten[keyName] = true
+		}
+
+		for k := range set {
+			if !blocksWritten[k] {
+				// Delete single key type
+				if diags := deleteSingleManagedKey(d, meta, keyType, k); diags != nil {
+					return diags
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func createUpdateManagedKeys(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	if _, ok := d.GetOk(consts.FieldAWS); ok {
+		if diags := writeManagedKeysData(d, meta, consts.FieldAWS); diags != nil {
+			return diags
+		}
+	}
+
+	if _, ok := d.GetOk(consts.FieldPKCS); ok {
+		if diags := writeManagedKeysData(d, meta, consts.FieldPKCS); diags != nil {
+			return diags
+		}
+	}
+
+	if _, ok := d.GetOk(consts.FieldAzure); ok {
+		if diags := writeManagedKeysData(d, meta, consts.FieldAzure); diags != nil {
+			return diags
+		}
+	}
+
+	// set ID to 'default'
+	d.SetId("default")
+
+	return readManagedKeys(ctx, d, meta)
+}
+
+func getRedactedFields(d *schema.ResourceData, providerType, name string, fields []string) map[string]interface{} {
+	m := map[string]interface{}{}
+
+	if mKeysBlocks, ok := d.GetOk(providerType); ok {
```

Comment:
Is there a specific reason why we need to iterate over all config resources for every `name`? I  wonder if we could fetch the  values directly from `d` e.g: `d.GetOk(providerType+"."+hash(name)+"."+field)` ?

Reply:
Ah that's a good point! I didn't think to use `hash(name)` to achieve that. I'll make the necessary changes

Reply:
Fixed in latest commit